### PR TITLE
Update memory allocation accounting for new pages and realloc.

### PR DESCRIPTION
**Pull request by agorrod**
They used to not always account for alignment overhead.

This is a prototype implementation - not yet ready for merging (or verified to fix the underlying issue).

### CODE REVIEW

File: src/btree/bt_slvg.c [hunk 1]
```diff
@@ -612,7 +613,10 @@ __slvg_trk_leaf(WT_SESSION_IMPL *session,
 		 * on every leaf page, and if you need to speed up the salvage,
 		 * it's probably a great place to start.
 		 */
-		WT_ERR(__wt_page_inmem(session, NULL, dsk, 0, &page));
+		WT_CLEAR(tmp);
+		tmp.data = dsk;
+		tmp.memsize = dsk->mem_size;
```
Comment (keithbostic) [inline src/btree/bt_slvg.c]: @agorrod, this scares me a lot -- setting the WT_ITEM's memsize, some code is going to look at that some time and think the WT_ITEM has associated memory. I'd go back to passing in the image and a length.

File: src/os_posix/os_alloc.c [hunk 2]
```diff
@@ -134,6 +134,13 @@ __wt_realloc_aligned(WT_SESSION_IMPL *session,
 		WT_ASSERT(session, bytes_to_allocate != 0);
 		WT_ASSERT(session, bytes_allocated < bytes_to_allocate);
 
+		/*
+		 * We are going to allocate an aligned buffer: round up to
+		 * the aligned size.
+		 */
+		bytes_to_allocate =
+		    WT_ALIGN(bytes_to_allocate, S2C(session)->buffer_alignment);
+
```
Comment (keithbostic) [inline src/os_posix/os_alloc.c]: @agorrod, this looks questionable to me: we don't need to increase the size of the buffer (necessarily) to reach a particular alignment. I agree that means we won't know the size of the buffer with as much accuracy, but it also means we don't waste so much space?

**Comment by michaelcahill** [comment]
Still to do in this branch:

1. explain why we round up the allocation size when calling `posix_memalign`; and
2. check that we can set `buffer_allocation = 0` if direct I/O is not configured on Linux.

**Comment by michaelcahill** [comment]
For the record, SERVER-17424 is the primary test case for this change.

**Comment by agorrod** [comment]
Thanks @michaelcahill lgtm - I'll merge.

**Pull request merged**